### PR TITLE
Vine Futures Executor

### PR DESCRIPTION
**Pull request by BarrySlyDelgado**

### CODE REVIEW

File: taskvine/src/manager/vine_manager.c [hunk 1]
```diff
@@ -2773,7 +2775,7 @@ static int vine_manager_check_inputs_available( struct vine_manager *q, struct v
 	LIST_ITERATE(t->input_mounts,m) {
 		struct vine_file *f = m->file;
 		if(f->type==VINE_TEMP) {
-			if(!vine_file_replica_table_exists_somewhere(q,f->cached_name)) {
+			if(!vine_file_replica_table_exists_somewhere(q,f->cached_name) &&  f->created) {
 				/* XXX we cannot tell between temp that was never run, and one that failed! */
```
Comment (dthain) [inline taskvine/src/manager/vine_manager.c:2779]: This comment is no longer true!

File: taskvine/src/bindings/python3/ndcctools/taskvine/manager.py [hunk 2]
```diff
@@ -1891,3 +1893,30 @@ def _write_config(self):
 
     def set_environment(self, env):
         self._env_file = env
+
```
Comment (dthain) [inline taskvine/src/bindings/python3/ndcctools/taskvine/manager.py:1896]: Perhaps the Executor should be in a separate module?

**Review by dthain — COMMENTED** [review]
This looks like it's on the right track.  Still needs some documentation, an example program, and a test case to be complete.

File: taskvine/src/examples/vine_example_future_matrix.py [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+                result[i][j] += a[i][k] * b[k][j]
+    return result
+
+futures = []
+
+levels = 3
+count = 0
+n = 400
+matricies = [generate_random_matrix(n) for x in range(2**levels)]
+print('Generated Matricies...')
```
Comment (dthain) [inline taskvine/src/examples/vine_example_future_matrix.py:69]: Matrices

File: doc/manuals/taskvine/index.md [hunk 4]
```diff
@@ -1423,6 +1423,68 @@ conda install -y -p my-env -c conda-forge conda-pack
 conda run -p my-env conda-pack
 ```
 
+### Futures
+
+TaskVine provides a futures executor model which is a subclass
+of pythons concurrent futures executor. A function along with its
```
Comment (dthain) [inline doc/manuals/taskvine/index.md:1429]: Python's

File: taskvine/src/examples/vine_example_future_matrix.py [hunk 5]
```diff
@@ -0,0 +1,106 @@
+#!/usr/bin/env python
+
+# Copyright (C) 2023- The University of Notre Dame
+# This software is distributed under the GNU General Public License.
+# See the file COPYING for details.
+
+# This example shows usage of TaskVine's future Executor
+# It performs matrix multiplications on multiple matricies
```
Comment (dthain) [inline taskvine/src/examples/vine_example_future_matrix.py:8]: throughout: matrices

File: taskvine/src/examples/vine_example_future_matrix.py [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+
+levels = 3
+count = 0
+n = 400
+matricies = [generate_random_matrix(n) for x in range(2**levels)]
+print('Generated Matricies...')
+
+executor = vine.Executor(manager_name='vine_matrix_example', batch_type='condor')
+executor.manager.enable_peer_transfers()
+executor.set("memory", 8000)
```
Comment (dthain) [inline taskvine/src/examples/vine_example_future_matrix.py:73]: This is a notable bit here: you are setting properties for all future tasks.  Highlight with some commentary.

File: taskvine/src/examples/vine_example_future_matrix.py [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+executor.set("disk", 8000)
+env_spec = {"conda": {"channels": ["conda-forge"],"packages": ["python","pip","conda","conda-pack","cloudpickle"]}}
```
Comment (dthain) [inline taskvine/src/examples/vine_example_future_matrix.py:75]: Explain this too

**Review by dthain — APPROVED** [review]

**Review by dthain — CHANGES_REQUESTED** [review]
Whoops, I meant to select 'request changes' before, see comments from prior review.

**Review by dthain — APPROVED** [review]

**Comment by dthain** [comment]
RTM?

**Comment by BarrySlyDelgado** [comment]
Yes.

**Pull request merged**